framework/util/marking_layers.h

```diff
@@ -0,0 +1,167 @@
+** AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+** LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+** FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+** DEALINGS IN THE SOFTWARE.
+*/
+
+#ifndef GFXRECON_UTIL_MARKING_LAYER_H
+#define GFXRECON_UTIL_MARKING_LAYER_H
+
+#include "vulkan/vulkan.h"
+#include "util/defines.h"
+#include "format/format.h"
+
+#include "decode/vulkan_object_info.h"
+#include "decode/vulkan_object_info_table.h"
+
+#include <string>
+#include <set>
+#include <unordered_map>
+
+GFXRECON_BEGIN_NAMESPACE(gfxrecon)
+GFXRECON_BEGIN_NAMESPACE(util)
+
+typedef void MarkInjectedCallback(void*);
+struct MarkInjectedCallbacks
+{
+    void*                 user_data;
+    MarkInjectedCallback* begin_injected = nullptr;
+    MarkInjectedCallback* end_injected   = nullptr;
+};
+
+class MarkingLayersUtil
+{
+  public:
+    static MarkingLayersUtil& instance()
+    {
+        static MarkingLayersUtil instance;
+        return instance;
+    }
+
+    template <typename T>
+    void BeginInjected(const decode::VulkanObjectInfo<T>* info, const std::string& file = "", int line = 0)
+    {
+        if constexpr (is_debug_)
+        {
+            if (!scope_closed_)
+            {
+                GFXRECON_LOG_FATAL("Marking was not closed, previous begin call at %s, %d", file_.c_str(), line_);
+                GFXRECON_ASSERT(false);
+            }
+            file_         = file;
+            line_         = line;
+            scope_closed_ = false;
+        }
+
+        callbacks_it_t entries = Callbacks(info);
+        if (entries == callbacks_.end())
+        {
+            return;
+        }
+
+        for (const MarkInjectedCallbacks& callback : entries->second)
+        {
+            callback.begin_injected(callback.user_data);
+        }
+    }
+
+    template <typename T>
+    void EndInjected(const decode::VulkanObjectInfo<T>* info)
+    {
+        if constexpr (is_debug_)
+        {
+            scope_closed_ = true;
+        }
+
+        MarkingLayersUtil::callbacks_it_t entries = Callbacks(info);
+        if (entries == callbacks_.end())
+        {
+            return;
+        }
+
+        for (const MarkInjectedCallbacks& callback : entries->second)
+        {
+            callback.end_injected(callback.user_data);
+        }
+    }
+
+    void AddCallbacks(format::HandleId physical_device_id, const VkPhysicalDeviceToolProperties& tool_properties)
+    {
+        // Do not add callbacks for the layers that were not enabled by the respective argument
+        if (layer_names_.count(tool_properties.layer) == 0)
+        {
+            return;
+        }
+        GFXRECON_LOG_DEBUG(
+            "Adding a %s layer callbacks from physical device %" PRIu64, tool_properties.layer, physical_device_id);
+        callbacks_[physical_device_id].emplace_back(*reinterpret_cast<MarkInjectedCallbacks*>(tool_properties.pNext));
```

Comment:
Where is it documented in the spec that a tools layer's VkPhysicalDeviceToolProperties::pNext is a callback?